### PR TITLE
Proposed fix to Issue #97: Move the Saml2AuthenticationModule to an own assembly

### PR DESCRIPTION
**Pull request by TobbeHolmstrom**
Removed the dependency on System.Web in the core library and from the Owin assembly. Added a new module for http.  

There might be an issue with code duplication between the MVC and HTTP modules if we do not introduce a utilities assembly or something like that as some replacement methods for System.Web are repeated in both projects.

The replacement for federation config section loading is a bit rough. I will continue to look at this.

### CODE REVIEW

**Comment by AndersAbel** [comment]
The nuspec file for the MVC project need to be updated to have a strict version reference to Kentor.AuthServices.HttpModule.

File: Kentor.AuthServices.Mvc/CommandResultExtensions.cs [hunk 1]
```diff
@@ -1,10 +1,14 @@
 ﻿using Kentor.AuthServices.WebSso;
 using System;
 using System.Collections.Generic;
+using System.Diagnostics.CodeAnalysis;
```
Comment (AndersAbel) [inline Kentor.AuthServices.Mvc/CommandResultExtensions.cs]: Is this really needed?

File: Kentor.AuthServices.Owin/KentorAuthServicesAuthenticationOptions.cs [hunk 2]
```diff
@@ -75,5 +81,23 @@ public string Caption
...
+
+        private void FederationIdentityConfiguration(FederationIdentityConfiguration identity)
+        {
+            federationIdentityConfiguration = identity;
+        }
+
+        /// <summary>
+        /// The federation identity configuration. 
+        /// </summary>
+        public IdentityConfiguration IdentityConfiguration
```
Comment (AndersAbel) [inline Kentor.AuthServices.Owin/KentorAuthServicesAuthenticationOptions.cs]: I want a better name, to make it clear that this relates to System.IdentityModel and not AuthServices' own federation element.

File: Kentor.AuthServices.Tests/HttpModule/CommandResultHttpTests.cs [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+            new CommandResult()
+            {
+                Cacheability = Cacheability.ServerAndNoCache
+            }.Apply(response);
+
+            cache.Received().SetCacheability(HttpCacheability.ServerAndNoCache);
+        }
+
+        [TestMethod]
+        public void CommandResult_Apply_HandleRedirect()
```
Comment (AndersAbel) [inline Kentor.AuthServices.Tests/HttpModule/CommandResultHttpTests.cs]: CommandResultHttp_ (and the rest of the file)

File: Kentor.AuthServices.sln [hunk 4]
```diff
@@ -1,8 +1,6 @@
 ﻿
 Microsoft Visual Studio Solution File, Format Version 12.00
-# Visual Studio 2013
```
Comment (AndersAbel) [inline Kentor.AuthServices.sln]: Remove these lines.

File: Kentor.AuthServices/Configuration/FederationIdentityConfiguration.cs [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace Kentor.AuthServices.Configuration
+{
+    /// <summary>
+    /// <c>FederationIdentityConfiguration</c> is used to load a FederationConfiguration directly from Web.Config without the need to use System.Web.
+    /// </summary>
+    public class FederationIdentityConfiguration : FederationConfiguration
```
Comment (AndersAbel) [inline Kentor.AuthServices/Configuration/FederationIdentityConfiguration.cs]: Maybe SystemIdentityModelConfigurationsection or something that indicates that this is from System.IdentityModel and not to be confused with AuthServices' own FederationElement

File: Kentor.AuthServices/Internal/PathHelper.cs [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace Kentor.AuthServices.Internal
+{
+    /// <summary>
+    /// Class to help with mapping virtual paths relative to the server.
+    /// </summary>
+    public static class PathHelper
```
Comment (AndersAbel) [inline Kentor.AuthServices/Internal/PathHelper.cs]: Make the class internal

File: Kentor.AuthServices/Internal/PathHelper.cs [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+            virtualPath = virtualPath.Replace('/', '\\');
+            return Path.Combine(BasePath, virtualPath);
+        }
+
+        /// <summary>
+        /// Determines if a virtual path is relative or not.
+        /// </summary>
+        /// <param name="virtualPath">The path that is to be tested.</param>
+        /// <returns>True if the path is relative otherwise false.</returns>
+        public static bool IsRelative(String virtualPath)
```
Comment (AndersAbel) [inline Kentor.AuthServices/Internal/PathHelper.cs]: Rename to IsWebRootRelative?

File: Kentor.AuthServices/Internal/QueryStringHelper.cs [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+    /// Class implements static methods to help parse a query string.
+    /// </summary>
+    public static class QueryStringHelper
+    {
+        /// <summary>
+        /// Splits a query string into its key/value pairs. 
+        /// </summary>
+        /// <param name="queryString">A query string, with or without the leading '?' character.</param>
+        /// <returns>A collecktion with the parsed keys and values.</returns>
+        public static NameValueCollection ParseQueryString(String queryString)
```
Comment (AndersAbel) [inline Kentor.AuthServices/Internal/QueryStringHelper.cs]: Is the old NameValueCollection really the best type to return? Could we go with a more modern IDictionary<string, string>? What is used in Owin? Could we use the same type?

File: Kentor.AuthServices/Internal/QueryStringHelper.cs [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+        }
+
+
+        /// <summary>
+        /// Encodes an url string and acts as a replacement for HttpUtility.UrlEncode.
+        /// </summary>
+        /// <param name="url">The url to be encoded.</param>
+        /// <returns>An encoded url.</returns>
+        [System.Diagnostics.CodeAnalysis.SuppressMessage("Microsoft.Design", "CA1054:UriParametersShouldNotBeStrings", MessageId = "0#", Justification="This is a drop-in replacement for a microsoft library function."), System.Diagnostics.CodeAnalysis.SuppressMessage("Microsoft.Design", "CA1055:UriReturnValuesShouldNotBeStrings")]
+        public static String UrlEncode(string url)
```
Comment (AndersAbel) [inline Kentor.AuthServices/Internal/QueryStringHelper.cs]: Check System.Net.WebUtility.UrlEncode instead.

File: Kentor.AuthServices/SAML2P/Saml2PSecurityTokenHandler.cs [hunk 10]
```diff
@@ -39,7 +39,7 @@ public Saml2PSecurityTokenHandler(EntityId spEntityId)
             {
                 IssuerNameRegistry = new ReturnRequestedIssuerNameRegistry(),
                 AudienceRestriction = audienceRestriction,
-                SaveBootstrapContext = FederatedAuthentication.FederationConfiguration.IdentityConfiguration.SaveBootstrapContext
+                SaveBootstrapContext = new FederationIdentityConfiguration().IdentityConfiguration.SaveBootstrapContext
```
Comment (AndersAbel) [inline Kentor.AuthServices/SAML2P/Saml2PSecurityTokenHandler.cs]: Not perfect, but I'm planning to ditch the Saml2PSecurityTokenHandler entirely soon, so let it be.

File: Kentor.AuthServices.Mvc/CommandResultExtensions.cs (hunk 1 above)
Comment (TobbeHolmstrom) [inline Kentor.AuthServices.Mvc/CommandResultExtensions.cs]: Seems it is used by ExcludeFromCodeCoverage, so I am thinking of leaving it in.

Comment (AndersAbel) [inline Kentor.AuthServices.Mvc/CommandResultExtensions.cs]: Then it's needed.

**Comment by AndersAbel** [comment]
The new HttpModule project needs to get Code Analysis and XML documentation export enabled. The latter is needed to get the warnings if public methods are not documented.

**Pull request merged**